modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/AbstractFreeList.java

```diff
@@ -267,6 +270,136 @@ private int addRowFragment(
+            int itemId,
+            IoStatisticsHolder statHolder
+        ) throws IgniteCheckedException {
+            DataPageIO io = (DataPageIO) iox;
+
+            Boolean updated = updateTtl(pageId, page, pageAddr, io, row, itemId, walPlc, statHolder);
+
+            evictionTracker.touchPage(pageId);
+
+            return updated;
+        }
+
+        /**
+         * @param pageId Page ID.
+         * @param page Page pointer.
+         * @param pageAddr Page address.
+         * @param io IO.
+         * @param row Row.
+         * @param itemId Item id.
+         * @return {@code true} If the expiration time was updated.
+         * @throws IgniteCheckedException If failed.
+         */
+        private Boolean updateTtl(
+            long pageId,
+            long page,
+            long pageAddr,
+            DataPageIO io,
+            T row,
+            int itemId,
+            Boolean walPlc,
+            IoStatisticsHolder statHolder
+        ) throws IgniteCheckedException {
+            // Read payload to determine that the row is fragmented.
+            DataPagePayload data = io.readPayload(pageAddr, itemId, pageMem.realPageSize(grpId));
+            if (data.nextLink() == 0) {
+                // The data is not fragmented, so it is one page update.
+                io.updateExpirationTime(pageAddr, data.offset(), (CacheDataRow) row);
+
+                return Boolean.TRUE;
+            }
+
+            // The data is fragmented, so need to update the required fragment.
+            long nextLink = 0L;
+            boolean pageUpdated = false;
+            int scannedBytes = 0;
+            int updatedBytes = 0;
+
+            // We under write lock here, so it is safe to read/write the first page.
+            boolean firstPage = true;
+
+            do {
+                final int cutItemId = firstPage ? itemId : PageIdUtils.itemId(nextLink);
+                final long curPageId = firstPage ? pageId : PageIdUtils.pageId(nextLink);
+
+                final long curPage = firstPage ? page : pageMem.acquirePage(grpId, curPageId, statHolder);
+
+                try {
+                    long curPageAddr = firstPage ? pageAddr : pageMem.writeLock(grpId, curPageId, curPage);
+
+                    try {
+                        assert curPageAddr != 0L : "Failed to acquire write lock on the page [grpId=" + grpId + ", pageId=" + curPageId +
+                            ", page=" + curPage + ", firstPage=" + firstPage + ", nextLink=" + nextLink + ']';
+
+                        ByteBuffer buf = pageMem.pageBuffer(curPageAddr);
+
+                        DataPageIO curIo = DataPageIO.VERSIONS.forPage(curPageAddr);
+
+                        data = curIo.readPayload(curPageAddr, cutItemId, pageMem.realPageSize(grpId));
+
+                        buf.position(data.offset());
+                        buf.limit(data.offset() + data.payloadSize());
+
+                        int res = curIo.updateExpirationTimeFragmentData(
```

Comment:
Please rename "res" into something better, or add a comment with a meaning

Reply:
Renamed to `updatedOnCurrPage`